packages/astro/src/integrations/hooks.ts

```diff
@@ -558,6 +559,47 @@ export async function runHookBuildDone({
 	}
 }
 
+export async function runHookRouteSetup({
+	route,
+	settings,
+	logger,
+}: {
+	route: RouteOptions;
+	settings: AstroSettings;
+	logger: Logger;
+}) {
+	const prerenderChangeLogs: { integrationName: string; value: boolean | undefined }[] = [];
+
+	for (const integration of settings.config.integrations) {
+		if (integration?.hooks?.['astro:route:setup']) {
```

Comment:
I think this doesn't actually run on every request now, only when picked up by the scanner. If it were to happen on every request I think we should avoid looping over every integration and instead collect the ones with this hook at an earlier stage.

Reply:
It actually runs once per route right now since Vite would cache it, but in the future with further refactoring, I think it may run per request again because at that point we need to decide which Vite environment to use before running Vite, but we can still certainly cache it then.

In any case, I think this hook function shouldn't be too heavy though since it only does checks if the hooks exists.